### DISCUSSION TITLE
Installing new cask throws Permission denied @ dir_initialize

### DISCUSSION THREAD

**Discussion by cesarcoatl** [Homebrew/discussions#862] [Everyday usage] [answered]
I am creating a cask formula for Sophos Home, and when testing the installation I get the following output:

```bash
$ brew install --cask sophos-home
==> Downloading https://s.home.sophos.com/download/SophosInstall.zip
Already downloaded: /Users/thecesrom/Library/Caches/Homebrew/downloads/addc4f47833905d902cd1c40fb5ce26538ca842e6c1322af5413cac48704a712--SophosInstall.zip
Warning: No checksum defined for cask 'sophos-home', skipping verification.
==> Installing Cask sophos-home
cp: /var/folders/2v/q2lv5pyx4w90bm8_lndfhwf80000gn/T/d20210218-68952-15bbn3k/Sophos Installer Components/.: Permission denied
==> Purging files for version latest of Cask sophos-home
Error: Permission denied @ dir_initialize - /var/folders/2v/q2lv5pyx4w90bm8_lndfhwf80000gn/T/d20210218-68952-15bbn3k/Sophos Installer Components
```

The ZIP file deflates into:

```
|____SophosInstall
| |____Sophos Installer.app
| |____Sophos Installer Components
```

❗  `Sophos Installer Components` directory is empty. 

And `Sophos Installer.app`:

```
|____Sophos Installer.app
| |____Contents
| | |_____CodeSignature
| | | |____CodeResources
| | |____Library
| | | |____LaunchServices
| | | | |____com.sophos.macendpoint.Installer.HelperTool
| | |____MacOS
| | | |____Sophos Home Installer
| | |____Resources
| | | |____...
| | |____CodeResources
| | |____Info.plist
| | |____PkgInfo
```

I am using `installer script` like this:

```ruby
  installer script: {
    executable: "#{staged_path}/SophosInstall/Sophos Installer.app/Contents/MacOS/Sophos Home Installer",
    sudo:       true,
  }
```

But it always fails.

Has anyone faced this?

Your assistance is much appreciated.

**Comment by SMillerDev** [comment]
ringing @Homebrew/cask for suggestions

**Comment by cesarcoatl** [comment]
I'm only venturing a guess here, but could it be that since the `Sophos Installer Components` directory is empty that may be causing the problem?

**Comment by vitorgalvao** [comment]
> creating a cask formula

Just “cask”.

> ```shell
> Error: Permission denied @ dir_initialize
> ```

I remember errors like that popping up in other instances. Some we fixed but I’m not sure we ever found the cause.

But if you’re making the cask to submit to the main repo, know that it won’t be accepted as is.

> ```ruby
>   installer script: {
>     executable: "#{staged_path}/SophosInstall/Sophos Installer.app/Contents/MacOS/Sophos Home Installer",
>     sudo:       true,
>   }
> ```

Because that will presumably invoke a GUI, which we don’t allow. `installer script` is to invoke scripts, not to start apps. What you’re looking for is [`installer manual:`](https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/installer.md#installer-manual). Though I just realised why we get that so often; we have a bad example in the documentation.

**Reply by cesarcoatl** [reply]
No matter which installer option I choose, the error remains the same.

**Comment by cesarcoatl** [comment, marked as answer]
I believe I may have found the cause.

There is a huge difference between double-clicking the ZIP file and running the `unzip` command.

### Double-clicking
```bash
$ cd ~/Downloads
$ cd SophosInstall
$ ls -al 
total 0
drwxr-xr-x@  4 thecesrom  staff   128 Feb 19 13:06 .
drwx------@ 59 thecesrom  staff  1888 Feb 19 13:04 ..
drwxrwxr-x@  2 thecesrom  staff    64 Jan  1  1980 Sophos Installer Components
drwxr-xr-x@  3 thecesrom  staff    96 Feb 19 13:04 Sophos Installer.app
```

### `unzip` command
```bash
$ cd ~/Downloads
$ unzip -qq -o -d SophosInstall SophosInstall.zip
$ cd SophosInstall
$ ls -al
total 0
drwxr-xr-x@  4 thecesrom  staff   128 Feb 19 13:09 .
drwx------@ 59 thecesrom  staff  1888 Feb 19 13:04 ..
d---------   2 thecesrom  staff    64 Jan  1  1980 Sophos Installer Components
drwxr-xr-x@  3 thecesrom  staff    96 Feb 19 13:04 Sophos Installer.app
```
A simple `chmod 775 Sophos\ Installer\ Components` fixes the permissions.
```bash
$ ls -al
total 0
drwxr-xr-x@  4 thecesrom  staff   128 Feb 19 13:09 .
drwx------@ 59 thecesrom  staff  1888 Feb 19 13:04 ..
drwxrwxr-x@  2 thecesrom  staff    64 Jan  1  1980 Sophos Installer Components
drwxr-xr-x@  3 thecesrom  staff    96 Feb 19 13:04 Sophos Installer.app
```

It seems like permissions are broken from source, but it begs the question why double-clicking differs from running `unzip`.

**Reply by vitorgalvao** [reply]
I think we use (used?) `ditto`, not `unzip`, but the result may be the same.

> it begs the question why double-clicking differs from running `unzip`.

Probably because whatever unarchiving utility you’re using (be it Apple’s default or another) is trying to be helpful and changing permissions.

> A simple `chmod 775 Sophos\ Installer\ Components` fixes the permissions.

A [`preflight` block](https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/flight.md) with `set_permissions` should do it.

**Reply by cesarcoatl** [reply]
Thanks for your suggestions, @vitorgalvao, but unfortunately adding `preflight` had no impact.

I just can't afford investing more time into this.
